### ISSUE TITLE
fluentd ooms too much

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#29411]
@jimmidyson any ideas? It's a pig.

**Comment by Random-Liu** [comment]
XRef https://github.com/kubernetes/kubernetes/issues/27630.

**Comment by edsiper** [comment]
did you faced any issue ?

**Comment by jimmidyson** [comment]
How much memory are you seeing it use? Can you share any resource usage stats? We don't face issues with it, although it may not be on the scale you're using it.

**Comment by a-robinson** [comment]
It gets bad under high load, in particular. The main issue is the buffers filling up due to not being able to keep up with the rate of logs being written to disk (and read in by the input plugin) and apparently not having a large enough memory limit to handle the number of full buffers that it ends up with. We could put a band-aid on the problem by buffering to disk instead of memory, but then we'll just fill up the disk instead.

The real fix is some combination of the following, in decreasing order of pleasantness:

1. See if we can buy any extra headroom by optimizing our configuration
1. #28720
1. #23782
1. Telling users to log less (which is obviously not a great fix)

**Comment by a-robinson** [comment]
To be clear, I don't mean that we need to buy extra memory headroom so much as extra CPU headroom. No matter how much memory is allocated, fluentd won't be able to do its job if it can't keep up with at least the average log rate on each node.

**Comment by edsiper** [comment]
would you please share your fluentd.conf/td-agent.conf ?, I would like to learn more about your use case and suggest improvements.

**Comment by bprashanth** [comment]
If the assertion is that this isn't fluent but our inability to accurately size the pod, maybe we should start by profiling fluent un-chained and see how it does? I wonder if there are other confiurations that might make it more efficient, perhaps tag team with docker's logging drivers so it isn't free form parsing? 

@edsiper thanks! the pod yaml is 
https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/fluentd-gcp/fluentd-gcp.yaml
and the config is https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/fluentd-gcp/fluentd-gcp-image

**Comment by a-robinson** [comment]
Profiling is definitely the right place to start here. I think what I'd assert is that we definitely have a pod-sizing problem and *likely* also have a fluentd problem. Profiling will help with both.

**Comment by bprashanth** [comment]
@gmarek can maybe handle this as part of his (node) scale testing? filling up an n1-standard-32 seems to generate logs too fast for fluentd to keep up. You could argue that fluentd is not part of kubernetes, but it affects cluster stability at the end.

**Comment by igorpeshansky** [comment]
Does the config in the fluentd container specify the number of threads? If not, it defaults to one, and if there is sufficient latency in the receiving service, it'll fall behind. Try setting num_threads to 8 in the config.

**Comment by gmarek** [comment]
The only way I may be able to help with this is to provide a repro if needed, but I guess everyone can start up a 100 containers logging stuff at full speed. I won't have time to do debugging of any kind though...

**Comment by bprashanth** [comment]
This is part of node scalabitliy, we need to figure out the cpu required by fluentd to support "normal" log levels of running 100 containers on a node without getting starved so much that it ooms (lots of containers on node > fluetd ooms > out of disk > node not ready > containers rescheduled on healthy node... and you lose the entier cluster). 

@igorpeshansky we only give fluentd 100m cpu, we need to profile it at expected load to figure out how much it might need. I'm not sure what "expected load" is in this context, unfortunately.

**Comment by gmarek** [comment]
Yeah - it'll be hard. The 'expected load' varies greatly depending on applications running on cluster. What's worse we probably should be able to handle applications running in full debug mode so spitting out A LOT of logs...

**Comment by bprashanth** [comment]
this is part of @matchstick's "we need to start profiling disruptive apps" idea

**Comment by tagomoris** [comment]
If fluentd's memory buffer is growing continuously without any errors of output plugins, the only reason of it is that output plugins takes too long time to write logs into destination.
Are you using Google Cloud Logging? Response time of cloud services are sometimes relatively slow.
As @igorpeshansky mentioned, it looks better to specify `num_threads 8`.

**Comment by bprashanth** [comment]
Yeah that's a good point about cloud being slow. Will bumping up the thread count actually help if it's constrained to 1/10th of a core via cpu limits? I suppose if each thread is going to handle a different tcp connection to the cloud and offload data in parallel

**Comment by igorpeshansky** [comment]
Correct, threads are needed to absorb the latency, so they will mostly be in the wait state. I'm not certain Ruby will throttle threads if they run over the limit, but I think Kubernetes might compensate...

**Comment by tagomoris** [comment]
Ruby doesn't throttle threads, but uses just one CPU core either way. It should do nothing problematic.
IMO it might be better to add `num_threads 4` or larger specification in default configuration for google cloud logging if there are more reports about it.

**Comment by a-robinson** [comment]
Agreed, adding threads should be the first change we try.

**Comment by igorpeshansky** [comment]
@tagomoris: Yes, I intend to set the default num_threads to 8 in the Google config. I'm not sure Kubernetes uses the same config, though, so it may need to be changed in the fluentd container config as well.

**Comment by a-robinson** [comment]
cc @piosz @fgrzadkowski

**Comment by rbuskens** [comment]
https://github.com/kubernetes/kubernetes/pull/29801 has been created to set num_threads to 8 by default.

**Comment by Ghazgkull** [comment]
Does anyone know if #29801 made it into v1.3.6?

**Comment by a-robinson** [comment]
It certainly doesn't look like it: https://github.com/kubernetes/kubernetes/tree/release-1.3/cluster/addons/fluentd-gcp/fluentd-gcp-image

**Comment by igorpeshansky** [comment]
According to https://github.com/kubernetes/kubernetes/commit/35fb50ba4c5b950f71d949d69f8fb4d846caa6c2, it is in v1.5.0-alpha.0, v1.4.0-beta.0, and v1.4.0-alpha.3, but not 1.3.6.

**Comment by kzk** [comment]
Just for a reference to people who have similar issues, at #32762, we partially identified that fluent-plugin-google-cloud output is using much memory.

**Comment by gmarek** [comment]
@crassirostris

**Comment by Crassirostris** [comment]
As @kzk pointed out, we were able to identify the main problem with the memory consumption. We were able to reproduce those frequent ooms too. 

Fixing fluentd plugin and switching to the file buffering (#32762) should put an end to this problem.

**Comment by piosz** [comment]
closing in favor of #32762
